metrics/src/common.rs

```diff
@@ -0,0 +1,235 @@
+use crate::data::AtomicWindowedHistogram;
+use metrics_util::StreamingIntegers;
+use quanta::Clock;
+use std::borrow::Cow;
+use std::ops::Deref;
+use std::sync::atomic::{AtomicI64, AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+/// Optimized metric name.
+///
+/// This can either be a [`&'static str`](str) or [`String`].
+pub type MetricName = Cow<'static, str>;
+
+/// A scope, or context, for a metric.
+///
+/// Not interacted with directly by end users, and only exposed due to a lack of trait method
+/// visbility controls.
+///
+/// See also: [Sink::scoped](crate::Sink::scoped).
+#[derive(PartialEq, Eq, Hash, Clone)]
+pub enum MetricScope {
+    Root,
+    Nested(Vec<String>),
+}
+
+impl MetricScope {
+    pub(crate) fn into_scoped(self, name: MetricName) -> String {
+        match self {
+            MetricScope::Root => name.to_string(),
+            MetricScope::Nested(mut parts) => {
+                if !name.is_empty() {
+                    parts.push(name.to_string());
+                }
+                parts.join(".")
+            }
+        }
+    }
+}
+
+pub(crate) type MetricScopeHandle = u64;
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub(crate) enum MetricKind {
+    Counter,
+    Gauge,
+    Histogram,
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub(crate) enum MetricIdentifier {
+    Unlabeled(MetricName, MetricScopeHandle, MetricKind),
+}
+
+#[derive(Debug)]
+enum ValueState {
+    Counter(AtomicU64),
+    Gauge(AtomicI64),
+    Histogram(AtomicWindowedHistogram),
+}
+
+#[derive(Debug)]
+pub(crate) enum ValueSnapshot {
+    Counter(u64),
+    Gauge(i64),
+    Histogram(StreamingIntegers),
+}
+
+#[derive(Clone, Debug)]
+/// Handle to the underlying measurement for a metric.
+pub(crate) struct MetricValue {
+    state: Arc<ValueState>,
+}
+
+impl MetricValue {
+    fn new(state: ValueState) -> Self {
+        MetricValue {
+            state: Arc::new(state),
+        }
+    }
+
+    pub(crate) fn new_counter() -> Self {
```

Comment:
miiiiiiiiiight make sense to put all these pub(crate) impls into their own `impl` block?

Reply:
I made all the inner methods regular `pub` since the type itself is `pub(crate)`, and I also dropped the `new_` prefix on some of the methods to make it a little cleaner.